rfd/0214-identifier-first-login.md

```diff
@@ -0,0 +1,193 @@
+  repeated string UserMatchers = 19 [(gogoproto.jsontag) = "user_matchers,omitempty"]
+}
+```
+
+```proto
+// OIDCConnectorSpecV3 is an OIDC connector specification.
+//
+// It specifies configuration for Open ID Connect compatible external
+// identity provider: https://openid.net/specs/openid-connect-core-1_0.html
+message OIDCConnectorSpecV3 {
+ ...
+  // UserMatchers is a set of RegEx patterns to narrow down which username(s) this auth connector should
+  // match for identifier-first login.
+  repeated string UserMatchers = 21 [(gogoproto.jsontag) = "user_matchers,omitempty"]
+}
+```
+
+```proto
+// GithubConnectorSpecV3 is a Github connector specification.
+message GithubConnectorSpecV3 {
+   ...
+  // UserMatchers is a set of RegEx patterns to narrow down which username(s) this auth connector should
+  // match for identifier-first login.
+  repeated string UserMatchers = 10 [(gogoproto.jsontag) = "user_matchers,omitempty"]
+}
+```
+
+### UX
+
+<img src="./assets/0214-identifier-first-flow.png" width="800" />
+
+Upon visiting the login page for the first time, if there are one or more auth connectors in the cluster
+with an explicitly defined `user_matchers` field, they will be prompted to enter a username (1a).
+If there are no auth connectors with an explicitly defined `user_matchers` field, the login page will
+be the same as it is now prior to the implementation of this feature (with all auth connectors listed),
+this is because it means that all auth connectors match all users, and thus prompting the user for
+their username is redundant and a pointless extra step.
+
+After a user enters their username, their username will be matched to the auth connector(s) relevant to
+them based on their `user_matchers` field, and those connectors will be displayed to them. If they are
+only matched to one auth connector, they will be taken directly to that identity provider to log in.
+Their username input will be stored in `localStorage`, so the next time they visit the login page, they
+won't need to enter their username again. If they are matched to no connectors, an error
+message will be displayed.
+
+If a remembered user only has one connector, they will be shown a single "Log in with <connector-name>"
+button to log in with that connector. If a remembered user has no connectors, which can happen if auth
+connector configs were changed to exclude that user, they will be directed back to the default initial
+login screen and their remembered username will be cleared from memory.
+
+### Security
+
+This is a relatively safe feature, as we are not exposing any sensitive information. The only security
+concern would be that an attacker could theoretically poke around and get an idea of which usernames
+are mapped to auth connectors. For example, if they see that `joe@foo.com` is mapped to auth connectors,
+then they'll know that that email is likely real and they may try to target it with a phishing
+or social engineering attack.
```

Comment:
> they'll know that that email is likely real

As you mentioned above, the value here is an arbitrary string. The attacker would know that the `foo.com` domain is likely real, but they won't have any idea whether `joe@foo.com` is real. 

Right?

Reply:
That's correct, unless the auth connector was mapped specifically to `joe@foo.com`